creusot-contracts/src/std/iter/filter.rs

```diff
@@ -0,0 +1,129 @@
+            // for all elements where the predicate evaluated to true, since we don't actually have access to the closure's return value directly,
+            // only what the postcondition says about it.
+            (forall<f : &mut F, i : _> f.postcondition_mut((i,), true) != f.postcondition_mut((i,), false))
+        }
+    }
+}
+
+/// Asserts that `f` has no precondition: any closure state can be called with any input value
+/// In a future release this restriction may be lifted or weakened
+#[open]
+#[predicate]
+pub fn no_precondition<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : F, i : A> f.precondition((i,)) }
+}
+
+/// Asserts that the captures of `f` are used immutably
+/// In a future release this restriction may be lifted or weakened
+#[open]
+#[predicate]
+pub fn immutable<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : F, g : F> f.unnest(g) ==> f == g }
+}
+
+/// The postcondition of `f` does not depend on prophecies storied in its in environment.
+/// In practice this means that the postcondition is not testing the equality of a mutable borrow
+#[open]
+#[predicate(prophetic)]
+pub fn plain<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : &mut F, g : &mut F, i :_, b :_> *f == *g && ^f == ^g ==> f.postcondition_mut((i,), b) == g.postcondition_mut((i,), b) }
+}
+
+/// Asserts that the postcondition of `f` is *precise*: that there are never two possible values matching the postcondition
+#[open]
+#[predicate]
+pub fn precise<A, F: FnMut(A) -> bool>(f: F) -> bool {
+    pearlite! { forall<f : &mut F, i : _> f.postcondition_mut((i,), true) != f.postcondition_mut((i,), false) }
+}
+
+impl<I, F> Iterator for Filter<I, F>
+where
+    I: Iterator,
+    F: FnMut(&I::Item) -> bool,
+{
+    #[open]
+    #[predicate(prophetic)]
+    fn completed(&mut self) -> bool {
+        pearlite! {
+        (exists<inner : &mut _> inv(inner) && *inner == self.iter() && ^inner == (^self).iter()
+            && inner.completed()) && (*self).func() == (^self).func() }
+    }
+
+    #[open]
+    #[predicate(prophetic)]
+    fn produces(self, visited: Seq<Self::Item>, succ: Self) -> bool {
+        pearlite! {
+            self.func().unnest(succ.func()) &&
+            // f here is a mapping from indices of `visited` to those of `s`, where `s` is the whole sequence produced by the underlying iterator
+            // Interestingly, Z3 guesses `f` quite readily but gives up *totally* on `s`. However, the addition of the final assertions on the correctness of the values
+            // blocks z3's guess for `f`.
+            exists<s : Seq<Self::Item>, f : Mapping<Int, Int>> self.iter().produces(s, succ.iter()) &&
+                // `f` is a monotone mapping
+                (forall<i : _, j :_ > 0 <= i && i <= j && j < visited.len() ==> 0 <= f.get(i) && f.get(i) <= f.get(j) && f.get(j) < s.len()) &&
```

Comment:
```suggestion
                (forall<i : _, j :_ > 0 <= i && i < j && j < visited.len() ==> 0 <= f.get(i) && f.get(i) < f.get(j) && f.get(j) < s.len()) &&
```

THIS is injectivity.